codec/src/main/java/io/netty/handler/codec/compression/ZstdEncoder.java

```diff
@@ -0,0 +1,193 @@
+
+import com.github.luben.zstd.Zstd;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.EncoderException;
+import io.netty.handler.codec.MessageToByteEncoder;
+import io.netty.util.internal.ObjectUtil;
+import java.nio.ByteBuffer;
+
+import static io.netty.handler.codec.compression.ZstdConstants.*;
+
+/**
+ *  Compresses a {@link ByteBuf} using the Zstandard algorithm.
+ *  See <a href="https://facebook.github.io/zstd">Zstandard</a>.
+ */
+public final class ZstdEncoder extends MessageToByteEncoder<ByteBuf> {
+
+    private final int blockSize;
+    private final int compressionLevel;
+    private ByteBuf buffer;
+    private final int maxEncodeSize;
+
+    /**
+     * Creates a new Zstd encoder.
+     *
+     * Please note that if you use the default constructor, the default BLOCK_SIZE and MAX_BLOCK_SIZE
+     * will be used. If you want to specify BLOCK_SIZE and MAX_BLOCK_SIZE yourself,
+     * please use {@link ZstdEncoder(int,int)} constructor
+     */
+    public ZstdEncoder() {
+        this(DEFAULT_COMPRESSION_LEVEL, DEFAULT_BLOCK_SIZE, MAX_BLOCK_SIZE);
+    }
+
+    /**
+     * Creates a new Zstd encoder.
+     *  @param  compressionLevel
+     *            specifies the level of the compression
+     */
+    public ZstdEncoder(int compressionLevel) {
+        this(compressionLevel, DEFAULT_BLOCK_SIZE, MAX_BLOCK_SIZE);
+    }
+
+    /**
+     * Creates a new Zstd encoder.
+     *  @param  blockSize
+     *            is used to calculate the compressionLevel
+     *  @param  maxEncodeSize
+     *            specifies the size of the largest compressed object
+     */
+    public ZstdEncoder(int blockSize, int maxEncodeSize) {
+        this(DEFAULT_COMPRESSION_LEVEL, blockSize, maxEncodeSize);
+    }
+
+    /**
+     * @param  blockSize
+     *           is used to calculate the compressionLevel
+     * @param  maxEncodeSize
+     *           specifies the size of the largest compressed object
+     * @param  compressionLevel
+     *           specifies the level of the compression
+     */
+    private ZstdEncoder(int compressionLevel, int blockSize, int maxEncodeSize) {
+        super(true);
+        this.compressionLevel = ObjectUtil.checkPositive(compressionLevel, "compressionLevel");
+        this.blockSize = blockSize;
+        this.maxEncodeSize = ObjectUtil.checkPositive(maxEncodeSize, "maxEncodeSize");
+    }
+
+    @Override
+    protected ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect) {
+        return allocateByteBuffer(ctx, msg, preferDirect);
+    }
+
+    private ByteBuf allocateByteBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect) {
+        if (!preferDirect) {
+            throw new EncoderException("dstBuff must be a direct buffer");
+        }
+
+        if (buffer == null) {
+            buffer = ctx.alloc().directBuffer(blockSize);
```

Comment:
imho it is very strange that this method returns a `ByteBuf` but also allocate another buffer as a side-effect.

Reply:
I think it would be better to do this somewhere different

Reply:
Thanks for review, how about making an assertion here